test/Transactions/Orleans.Transactions.Tests/Grains/MultiStateTransactionalGrains.cs

```diff
@@ -0,0 +1,144 @@
+                  loggerFactory)
+        {
+        }
+    }
+
+    public class DoubleStateTransactionalGrain : MultiStateTransactionalGrainBaseClass
+    {
+        public DoubleStateTransactionalGrain(
+            [TransactionalState("data1", TransactionTestConstants.TransactionStore)]
+            ITransactionalState<GrainData> data1,
+            [TransactionalState("data2", TransactionTestConstants.TransactionStore)]
+            ITransactionalState<GrainData> data2,
+            ILoggerFactory loggerFactory)
+            : base(new ITransactionalState<GrainData>[2] { data1, data2 }, loggerFactory)
+        {
+        }
+    }
+
+    public class SingleStateTransactionalGrain : MultiStateTransactionalGrainBaseClass
+    {
+        public SingleStateTransactionalGrain(
+            [TransactionalState("data", TransactionTestConstants.TransactionStore)] ITransactionalState<GrainData> data,
+            ILoggerFactory loggerFactory)
+            :base(new ITransactionalState<GrainData>[1]{data}, loggerFactory)
+        {
+        }
+    }
+
+    public class MultiStateTransactionalGrainBaseClass: Grain, ITransactionTestGrain
+    {
+        protected ITransactionalState<GrainData>[] dataArray;
+        private readonly ILoggerFactory loggerFactory;
+        protected ILogger logger;
+
+        public MultiStateTransactionalGrainBaseClass(
+            ITransactionalState<GrainData>[] dataArray,
+            ILoggerFactory loggerFactory)
+        {
+            this.dataArray = dataArray;
+            this.loggerFactory = loggerFactory;
+        }
+
+        public override Task OnActivateAsync()
+        {
+            this.logger = this.loggerFactory.CreateLogger(this.GetGrainIdentity().ToString());
+            return base.OnActivateAsync();
+        }
+
+        public Task Set(int newValue)
+        {
+            foreach (var data in this.dataArray)
+            {
+                TransactionalGrainUtils.Set(data, newValue, this.logger);
+            }
+            return Task.CompletedTask;
+        }
+
+        public Task<int[]> Add(int numberToAdd)
+        {
+            return Task.FromResult(this.dataArray.Select(data => TransactionalGrainUtils.Add(data, numberToAdd, this.logger)).ToArray());
+        }
+
+        public Task<int[]> Get()
+        {
+            return Task.FromResult(this.dataArray.Select(data => TransactionalGrainUtils.Get(data, this.logger)).ToArray());
+        }
+
+        public Task AddAndThrow(int numberToAdd)
+        {
+            foreach (var data in dataArray)
+            {
+                TransactionalGrainUtils.Add(data, numberToAdd, this.logger);
+            }
+            throw new Exception($"{GetType().Name} test exception");
+        }
+
+        public Task Deactivate()
+        {
+            DeactivateOnIdle();
+            return Task.CompletedTask;
+        }
+    }
+
+    public static class TransactionalGrainUtils
```

Comment:
Unclear why we need a utility class.  It doesn't seem to buy us anything and introduces ambiguity, as 'utility' is very unclear as to what it's supposed to do.
In general I discourage the use of 'utility' or 'helper' classes because their ambiguous nature allows them to become catchalls for random functions, harming clarity and maintainability.

Reply:
it can be merged back to the transactionalGrainBaseClass.